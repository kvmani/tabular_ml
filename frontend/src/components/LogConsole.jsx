import React, { useEffect, useMemo, useRef, useState } from 'react';
import useLogStream from '../hooks/useLogStream.js';

const LEVEL_METADATA = {
  ERROR: { label: 'Error', icon: '‚ùå', className: 'log-error' },
  WARNING: { label: 'Warning', icon: '‚ö†Ô∏è', className: 'log-warn' },
  INFO: { label: 'Info', icon: '‚ÑπÔ∏è', className: 'log-info' },
  DEBUG: { label: 'Debug', icon: 'üêû', className: 'log-debug' }
};

const LEVEL_ORDER = ['ERROR', 'WARNING', 'INFO', 'DEBUG'];

function formatTime(value) {
  try {
    const parsed = new Date(value);
    if (!Number.isNaN(parsed.getTime())) {
      return parsed.toLocaleTimeString([], { hour12: false });
    }
  } catch (error) {
    // Ignore and fall back to raw value.
  }
  return value;
}

export default function LogConsole({ stream }) {
  const logStream = stream ?? useLogStream();
  const {
    logs,
    connectionState,
    lastError,
    lastConnectionError,
    backlogWarning,
    isPaused,
    autoScroll,
    setAutoScroll,
    includeDebug,
    setIncludeDebug,
    clear,
    pause,
    resume,
    acknowledgeBacklogWarning
  } = logStream;

  const [levelFilters, setLevelFilters] = useState(() => ({
    ERROR: true,
    WARNING: true,
    INFO: true,
    DEBUG: includeDebug
  }));

  useEffect(() => {
    setLevelFilters((prev) => ({
      ...prev,
      DEBUG: includeDebug ? prev.DEBUG || true : false
    }));
  }, [includeDebug]);

  const visibleLogs = useMemo(
    () => logs.filter((entry) => levelFilters[entry.level] !== false),
    [logs, levelFilters]
  );

  const bodyRef = useRef(null);

  useEffect(() => {
    if (!autoScroll) {
      return;
    }
    const node = bodyRef.current;
    if (node) {
      node.scrollTop = node.scrollHeight;
    }
  }, [visibleLogs, autoScroll]);

  const handleLevelToggle = (level) => {
    setLevelFilters((prev) => {
      const next = { ...prev, [level]: !prev[level] };
      if (level === 'DEBUG') {
        setIncludeDebug(next[level]);
      }
      return next;
    });
  };

  const handlePauseToggle = () => {
    if (isPaused) {
      resume();
    } else {
      pause();
    }
  };

  const handleAutoScrollToggle = () => {
    setAutoScroll(!autoScroll);
  };

  const handleBacklogDismiss = () => {
    if (acknowledgeBacklogWarning) {
      acknowledgeBacklogWarning();
    }
  };

  return (
    <div className="log-console">
      <div className="log-console__header">
        <div className="log-console__title">
          <h3>Live Logs</h3>
          <span className={`log-console__status log-console__status--${connectionState}`}>
            {connectionState === 'open' ? 'Connected' : connectionState}
          </span>
        </div>
        <div className="log-console__actions">
          <button type="button" onClick={handlePauseToggle} className="log-console__button">
            {isPaused ? 'Resume' : 'Pause'}
          </button>
          <button type="button" onClick={clear} className="log-console__button">
            Clear
          </button>
        </div>
      </div>
      {lastConnectionError && (
        <div className="log-console__notice log-console__notice--connection" role="alert">
          <span aria-hidden="true">üîå</span>
          <div>
            <strong>Connection issue:</strong> {lastConnectionError}
          </div>
        </div>
      )}
      {backlogWarning && (
        <div className="log-console__notice log-console__notice--warning" role="status">
          <span aria-hidden="true">‚ö†Ô∏è</span>
          <div>
            <strong>Backpressure detected:</strong> {backlogWarning.message}
            <div className="log-console__notice-actions">
              <button type="button" className="log-console__button" onClick={handleBacklogDismiss}>
                Dismiss
              </button>
            </div>
          </div>
        </div>
      )}
      <div className="log-console__controls">
        <fieldset className="log-console__levels">
          <legend className="sr-only">Severity filters</legend>
          {LEVEL_ORDER.map((level) => {
            const meta = LEVEL_METADATA[level];
            return (
              <label key={level} className="log-console__level">
                <input
                  type="checkbox"
                  checked={Boolean(levelFilters[level])}
                  onChange={() => handleLevelToggle(level)}
                />
                <span>
                  {meta.icon} {meta.label}
                </span>
              </label>
            );
          })}
        </fieldset>
        <label className="log-console__level">
          <input type="checkbox" checked={autoScroll} onChange={handleAutoScrollToggle} />
          <span>Auto-scroll</span>
        </label>
      </div>
      <div
        className="log-console__body"
        role="log"
        aria-live="polite"
        aria-relevant="additions"
        tabIndex={0}
        ref={bodyRef}
      >
        {visibleLogs.length === 0 ? (
          <div className="log-console__empty">Logs will appear here as the system runs.</div>
        ) : (
          visibleLogs.map((entry) => {
            const meta = LEVEL_METADATA[entry.level] || LEVEL_METADATA.INFO;
            return (
              <div key={entry.id} className={`log-entry ${meta.className}`}>
                <div className="log-entry__icon" aria-hidden="true">
                  {meta.icon}
                </div>
                <div className="log-entry__content">
                  <div className="log-entry__meta">
                    <time dateTime={entry.timestamp}>{formatTime(entry.timestamp)}</time>
                    {entry.logger && <span className="log-entry__logger">{entry.logger}</span>}
                    {entry.module && <span className="log-entry__module">{entry.module}</span>}
                  </div>
                  <div className="log-entry__message">{entry.message}</div>
                </div>
              </div>
            );
          })
        )}
      </div>
      {lastError && (
        <div className="log-console__footer" role="alert">
          {lastError}
        </div>
      )}
    </div>
  );
}
