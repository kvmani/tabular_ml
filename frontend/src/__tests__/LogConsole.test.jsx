import React from 'react';
import { render } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import LogConsole from '../components/LogConsole.jsx';

const baseStream = {
  logs: [
    {
      id: '1',
      level: 'INFO',
      message: 'Application boot complete',
      timestamp: '2024-01-01T00:00:00Z',
      logger: 'backend.main',
      module: 'main'
    },
    {
      id: '2',
      level: 'WARNING',
      message: 'Cache miss for dataset preview',
      timestamp: '2024-01-01T00:00:01Z',
      logger: 'services.cache',
      module: 'cache'
    },
    {
      id: '3',
      level: 'ERROR',
      message: 'Failed to load dataset metadata',
      timestamp: '2024-01-01T00:00:02Z',
      logger: 'services.data',
      module: 'data_manager'
    },
    {
      id: '4',
      level: 'DEBUG',
      message: 'Trainer heartbeat received',
      timestamp: '2024-01-01T00:00:03Z',
      logger: 'services.training',
      module: 'trainer'
    }
  ],
  connectionState: 'open',
  lastError: null,
  isPaused: false,
  autoScroll: true,
  includeDebug: true,
  setAutoScroll: vi.fn(),
  setIncludeDebug: vi.fn(),
  clear: vi.fn(),
  pause: vi.fn(),
  resume: vi.fn()
};

describe('LogConsole', () => {
  it('renders mixed severity entries', () => {
    const { container } = render(<LogConsole stream={baseStream} />);
    expect(container).toMatchSnapshot();
  });
});
