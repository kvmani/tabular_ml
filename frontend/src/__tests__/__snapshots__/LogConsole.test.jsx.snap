// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LogConsole > renders mixed severity entries 1`] = `
<div>
  <div
    class="log-console"
  >
    <div
      class="log-console__header"
    >
      <div
        class="log-console__title"
      >
        <h3>
          Live Logs
        </h3>
        <span
          class="log-console__status log-console__status--open"
        >
          Connected
        </span>
      </div>
      <div
        class="log-console__actions"
      >
        <button
          class="log-console__button"
          type="button"
        >
          Pause
        </button>
        <button
          class="log-console__button"
          type="button"
        >
          Clear
        </button>
      </div>
    </div>
    <div
      class="log-console__controls"
    >
      <fieldset
        class="log-console__levels"
      >
        <legend
          class="sr-only"
        >
          Severity filters
        </legend>
        <label
          class="log-console__level"
        >
          <input
            checked=""
            type="checkbox"
          />
          <span>
            ‚ùå
             
            Error
          </span>
        </label>
        <label
          class="log-console__level"
        >
          <input
            checked=""
            type="checkbox"
          />
          <span>
            ‚ö†Ô∏è
             
            Warning
          </span>
        </label>
        <label
          class="log-console__level"
        >
          <input
            checked=""
            type="checkbox"
          />
          <span>
            ‚ÑπÔ∏è
             
            Info
          </span>
        </label>
        <label
          class="log-console__level"
        >
          <input
            checked=""
            type="checkbox"
          />
          <span>
            üêû
             
            Debug
          </span>
        </label>
      </fieldset>
      <label
        class="log-console__level"
      >
        <input
          checked=""
          type="checkbox"
        />
        <span>
          Auto-scroll
        </span>
      </label>
    </div>
    <div
      aria-live="polite"
      aria-relevant="additions"
      class="log-console__body"
      role="log"
      tabindex="0"
    >
      <div
        class="log-entry log-info"
      >
        <div
          aria-hidden="true"
          class="log-entry__icon"
        >
          ‚ÑπÔ∏è
        </div>
        <div
          class="log-entry__content"
        >
          <div
            class="log-entry__meta"
          >
            <time
              datetime="2024-01-01T00:00:00Z"
            >
              24:00:00
            </time>
            <span
              class="log-entry__logger"
            >
              backend.main
            </span>
            <span
              class="log-entry__module"
            >
              main
            </span>
          </div>
          <div
            class="log-entry__message"
          >
            Application boot complete
          </div>
        </div>
      </div>
      <div
        class="log-entry log-warn"
      >
        <div
          aria-hidden="true"
          class="log-entry__icon"
        >
          ‚ö†Ô∏è
        </div>
        <div
          class="log-entry__content"
        >
          <div
            class="log-entry__meta"
          >
            <time
              datetime="2024-01-01T00:00:01Z"
            >
              24:00:01
            </time>
            <span
              class="log-entry__logger"
            >
              services.cache
            </span>
            <span
              class="log-entry__module"
            >
              cache
            </span>
          </div>
          <div
            class="log-entry__message"
          >
            Cache miss for dataset preview
          </div>
        </div>
      </div>
      <div
        class="log-entry log-error"
      >
        <div
          aria-hidden="true"
          class="log-entry__icon"
        >
          ‚ùå
        </div>
        <div
          class="log-entry__content"
        >
          <div
            class="log-entry__meta"
          >
            <time
              datetime="2024-01-01T00:00:02Z"
            >
              24:00:02
            </time>
            <span
              class="log-entry__logger"
            >
              services.data
            </span>
            <span
              class="log-entry__module"
            >
              data_manager
            </span>
          </div>
          <div
            class="log-entry__message"
          >
            Failed to load dataset metadata
          </div>
        </div>
      </div>
      <div
        class="log-entry log-debug"
      >
        <div
          aria-hidden="true"
          class="log-entry__icon"
        >
          üêû
        </div>
        <div
          class="log-entry__content"
        >
          <div
            class="log-entry__meta"
          >
            <time
              datetime="2024-01-01T00:00:03Z"
            >
              24:00:03
            </time>
            <span
              class="log-entry__logger"
            >
              services.training
            </span>
            <span
              class="log-entry__module"
            >
              trainer
            </span>
          </div>
          <div
            class="log-entry__message"
          >
            Trainer heartbeat received
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
